### PR TITLE
8266349: Pass down requested page size to reserve_memory_special

### PR DESCRIPTION
**Pull request by kstefanj**
Please review this enhancement that passes the requested page size in `ReservedSpace` down to the OS-layer.

**Summary**
After recent refactoring `ReservedSpace` now store the page size used for a reservation. To make this more consistent this change passes down the page size from the `ReservedSpace` down to the OS-layer when using explicit large pages. Currently this is not really needed because we only support one large page size during the lifetime of the JVM, but this will change with PR #1153. That PR will be simpler to implement if the requested page size is passed down from the layers above. 

**Testing**
Local testing together with parts of PR #1153. Mach5 tier 1-3

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266349](https://bugs.openjdk.java.net/browse/JDK-8266349): Pass down requested page size to reserve_memory_special

### Reviewers
 * [Thomas Stuefe](https://openjdk.java.net/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to fc36939e96db796060ac662ed696c410c623238a
 * [Marcus G K Williams](https://openjdk.java.net/census#mgkwill) (@mgkwill - Author) ⚠️ Review applies to fc36939e96db796060ac662ed696c410c623238a

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3878/head:pull/3878` \
`$ git checkout pull/3878`

Update a local copy of the PR: \
`$ git checkout pull/3878` \
`$ git pull https://git.openjdk.java.net/jdk pull/3878/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3878`

View PR using the GUI difftool: \
`$ git pr show -t 3878`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3878.diff">https://git.openjdk.java.net/jdk/pull/3878.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/os/linux/os_linux.cpp [hunk 1]
```diff
@@ -3957,14 +3957,15 @@ bool os::Linux::commit_memory_special(size_t bytes,
...
                                                    bool exec) {
   assert(UseLargePages && UseHugeTLBFS, "only for Huge TLBFS large pages");
   assert(is_aligned(req_addr, alignment), "Must be");
-  assert(is_aligned(req_addr, os::large_page_size()), "Must be");
+  assert(is_aligned(req_addr, page_size), "Must be");
   assert(is_aligned(alignment, os::vm_allocation_granularity()), "Must be");
-  assert(is_power_of_2(os::large_page_size()), "Must be");
-  assert(bytes >= os::large_page_size(), "Shouldn't allocate large pages for small sizes");
+  assert(is_power_of_2(page_size), "Must be");
+  assert(bytes >= page_size, "Shouldn't allocate large pages for small sizes");
```
Comment (tstuefe) [inline src/hotspot/os/linux/os_linux.cpp:3968]: We could replace these two asserts with _page_sizes.contains.

File: src/hotspot/share/memory/virtualspace.cpp [hunk 2]
```diff
@@ -169,14 +169,14 @@ static char* reserve_memory(char* requested_address, const size_t size,
 }
 
 static char* reserve_memory_special(char* requested_address, const size_t size,
-                                    const size_t alignment,  bool exec) {
+                                    const size_t alignment, const size_t page_size, bool exec) {
```
Comment (tstuefe) [inline src/hotspot/share/memory/virtualspace.cpp:172]: I don't think the const makes sense... nor for the other arguments (but when in Rome...)

**Review by tstuefe — APPROVED** [review]
Very good. 

This is not only cosmetic but functional, no? Since before we would always os::large_page_size(); now, in the case of ReservedSpace::ReservedSpace(size), we use whatever os::page_size_for_region wants us to use. Lets see if any errors creep up.

..Thomas

File: src/hotspot/os/linux/os_linux.cpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/os/linux/os_linux.cpp:3968]: .. Well, the first one at least.

Comment (kstefanj) [reply src/hotspot/os/linux/os_linux.cpp:3968]: Was just about to say just that :) The second one should trigger on the case you mention, if we ever tried to use large pages for too small sizes.

File: src/hotspot/share/memory/virtualspace.cpp (hunk 2 above)
Comment (kstefanj) [reply src/hotspot/share/memory/virtualspace.cpp:172]: Yes, I'm staying in Rome :) But we could certainly clean it up later.

**Review by kstefanj — COMMENTED** [review]
> Very good.
> 

Thanks =)

> This is not only cosmetic but functional, no? Since before we would always os::large_page_size(); now, in the case of ReservedSpace::ReservedSpace(size), we use whatever os::page_size_for_region wants us to use. Lets see if any errors creep up.
> 

I think this should not change functionality. Before I started my refactoring we did not have `ReservedSpace::ReservedSpace(size)` but this function below where the second parameter was optional. As you can see, if no `preferred_page_size` was given we would use `os::page_size_for_region_unaligned(size, 1)` which is what now is done in `ReservedSpace::ReservedSpace(size)`:
```
ReservedSpace::ReservedSpace(size_t size, size_t preferred_page_size) : _fd_for_heap(-1) {
  bool has_preferred_page_size = preferred_page_size != 0;
  // Want to use large pages where possible and pad with small pages.
  size_t page_size = has_preferred_page_size ? preferred_page_size : os::page_size_for_region_unaligned(size, 1);
  bool large_pages = page_size != (size_t)os::vm_page_size();
  ...
```

And if we end up with a page size that isn't large we will never get down into `reserve_memory_special(...)`. Does this sound correct, or did I miss your point?

**Comment by tstuefe** [comment]
> > Very good.
> 
> Thanks =)
> 
> > This is not only cosmetic but functional, no? Since before we would always os::large_page_size(); now, in the case of ReservedSpace::ReservedSpace(size), we use whatever os::page_size_for_region wants us to use. Lets see if any errors creep up.
> 
> I think this should not change functionality. Before I started my refactoring we did not have `ReservedSpace::ReservedSpace(size)` but this function below where the second parameter was optional. As you can see, if no `preferred_page_size` was given we would use `os::page_size_for_region_unaligned(size, 1)` which is what now is done in `ReservedSpace::ReservedSpace(size)`:
> 
> ```
> ReservedSpace::ReservedSpace(size_t size, size_t preferred_page_size) : _fd_for_heap(-1) {
>   bool has_preferred_page_size = preferred_page_size != 0;
>   // Want to use large pages where possible and pad with small pages.
>   size_t page_size = has_preferred_page_size ? preferred_page_size : os::page_size_for_region_unaligned(size, 1);
>   bool large_pages = page_size != (size_t)os::vm_page_size();
>   ...
> ```
> 
> And if we end up with a page size that isn't large we will never get down into `reserve_memory_special(...)`. Does this sound correct, or did I miss your point?

No, this sounds right.

**Review by mgkwill — APPROVED** [review]
:+1:

**Comment by kstefanj** [comment]
Thanks @tstuefe and @mgkwill for reviewing.

/integrate

**Pull request closed without merging**